js-pkg/apps/demos/src/pages/crdt.tsx

```diff
@@ -0,0 +1,103 @@
+import { DRIFTDB_URL } from '../config'
+import { DbConnection } from 'driftdb'
+import { DriftDBProvider, RoomQRCode, StatusIndicator, useDatabase } from 'driftdb-react'
+import Head from 'next/head'
+import { useCallback, useRef, useState } from 'react'
+import * as Y from 'yjs'
+import 'codemirror/lib/codemirror.css'
+import type { Editor } from 'codemirror'
+
+class DriftYjsProvider {
+  constructor(private db: DbConnection, key: string, public doc: Y.Doc, onUpdate: () => void) {
+    let lastSeq = 0
+
+    this.db.subscribe(key, (data) => {
+      Y.applyUpdate(doc, data.value as Uint8Array)
+      lastSeq = data.seq
+    }, (size: number) => {
+      if (size > 30) {
+        // compact
+        let update = Y.encodeStateAsUpdate(doc)
+        this.db.send({
+          type: 'push',
+          key: key,
+          value: update,
+          action: { type: 'compact', seq: lastSeq }
+        })
+      }
+    })
+
+    doc.on('update', (update: Uint8Array) => {
+      this.db.send({
+        type: 'push',
+        key: key,
+        value: update,
+        action: { type: 'append' }
+      })
+      onUpdate()
+    })
+  }
+}
+
+function useYDoc(key: string) {
+  const [_, setVersion] = useState(0)
+  const db = useDatabase()
+  const provider = useRef<DriftYjsProvider | null>(null)
+  if (provider.current == null) {
+    const doc = new Y.Doc()
+    provider.current = new DriftYjsProvider(db, key, doc, () => setVersion((i) => i + 1))
+  }
+
+  return provider.current.doc
+}
+
+function CrdtDemo() {
+  const ydoc = useYDoc('text')
+  const editorRef = useRef<Editor | null>(null)
+
+  const codeMirrorRef = useCallback((ref: HTMLTextAreaElement | null) => {
+    if (ref == null) {
+      if (editorRef.current != null) {
+        (editorRef.current as any).toTextArea()
+        editorRef.current = null
```

Comment:
What does the `toTextArea` method do? (I couldn't find any docs on it on the CodeMirror site.)

Reply:
I think it came from [this post](https://stackoverflow.com/a/18829236), but you're right that it's suspicious I can't find any other source of it.

Reply:
Ah, here it is:

https://codemirror.net/5/doc/manual.html

It looks like the reason it needs to be coerced to `any` is that `toTextArea` only exists on a `CodeMirror` instance that was created by `fromTextArea`.